File: nodejs/test/e2e/rpc_server_misc.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { randomUUID } from "node:crypto";
import { mkdirSync, rmSync } from "node:fs";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import { approveAll, CopilotClient, RuntimeConnection } from "../../src/index.js";
import { createSdkTestContext, DEFAULT_GITHUB_TOKEN } from "./harness/sdkTestContext.js";
import { formatError, waitForCondition } from "./harness/sdkTestHelper.js";

describe("Miscellaneous server-scoped RPC", async () => {
    const { copilotClient: client, env, workDir } = await createSdkTestContext();

    function createUniqueDirectory(prefix: string): string {
        const directory = join(workDir, `${prefix}-${randomUUID()}`);
        mkdirSync(directory, { recursive: true });
        return directory;
    }

    function createClient(extraEnv: Record<string, string | undefined> = {}): CopilotClient {
        return new CopilotClient({
            workingDirectory: workDir,
            env: {
                ...env,
                ...extraEnv,
            },
            logLevel: "error",
            connection: RuntimeConnection.forStdio({ path: process.env.COPILOT_CLI_PATH }),
            gitHubToken: DEFAULT_GITHUB_TOKEN,
        });
    }

    async function createIsolatedStartedClient(): Promise<{
        client: CopilotClient;
        home: string;
    }> {
        const home = createUniqueDirectory("copilot-e2e-misc-home");
        const isolatedClient = createClient({
            COPILOT_HOME: home,
            GH_CONFIG_DIR: home,
            XDG_CONFIG_HOME: home,
            XDG_STATE_HOME: home,
        });
        try {
            await isolatedClient.start();
            return { client: isolatedClient, home };
        } catch (error) {
            await disposeIsolated(isolatedClient, home);
            throw error;
        }
    }

    async function disposeIsolated(isolatedClient: CopilotClient, home: string): Promise<void> {
        try {
            await isolatedClient.forceStop();
        } catch {
            // Best-effort cleanup.
        }
        tryRemoveDirectory(home);
    }

    async function forceStop(target: CopilotClient): Promise<void> {
        try {
            await target.forceStop();
        } catch {
            // Runtime may already be gone.
        }
    }

    function tryRemoveDirectory(directory: string): void {
        try {
            rmSync(directory, { recursive: true, force: true });
        } catch {
            // Temp directories are reclaimed by the harness/OS.
        }
    }

    it("should reload user settings", { timeout: 120_000 }, async () => {
        await client.start();

        await client.rpc.user.settings.reload();
    });

    it("should report agent registry spawn gate closed", { timeout: 120_000 }, async () => {
        const { client: isolatedClient, home } = await createIsolatedStartedClient();
        try {
            await expect(
                isolatedClient.rpc.agentRegistry.spawn({ cwd: workDir })
            ).rejects.toSatisfy((error: unknown) => {
                const message = formatError(error);
                expect(message.toLowerCase()).not.toContain("unhandled method");
                expect(message.toLowerCase()).toContain("agentregistry.spawn");
                expect(
                    message.toLowerCase().includes("not enabled") ||
                        message.toLowerCase().includes("no delegate")
                ).toBe(true);
                return true;
            });
        } finally {
            await disposeIsolated(isolatedClient, home);
        }
    });

    it("should shut down owned runtime", { timeout: 120_000 }, async () => {
        const dedicatedClient = createClient();
        try {
            await dedicatedClient.start();
            await dedicatedClient.rpc.user.settings.reload();

            await dedicatedClient.rpc.runtime.shutdown();

            await waitForCondition(
                async () => {
                    try {
                        await dedicatedClient.rpc.user.settings.reload();
                        return false;
                    } catch {
                        return true;
                    }
                },
                {
                    timeoutMs: 15_000,
                    intervalMs: 100,
                    timeoutMessage: "Runtime kept serving RPCs after a graceful shutdown.",
                }
            );
        } finally {
            await forceStop(dedicatedClient);
        }
    });

    it(
        "should report not found when opening session without context",
        { timeout: 120_000 },
        async () => {
            const { client: isolatedClient, home } = await createIsolatedStartedClient();
            try {
                const result = await isolatedClient.rpc.sessions.open({ kind: "resumeLast" });

                expect(result.status).toBe("not_found");
                expect(result.sessionId ?? null).toBeNull();
            } finally {
                await disposeIsolated(isolatedClient, home);
            }
        }
    );

    it(
        "should reject send attachments from non extension connection",
        { timeout: 120_000 },
        async () => {
            const session = await client.createSession({ onPermissionRequest: approveAll });
            try {
                await expect(
                    session.rpc.extensions.sendAttachmentsToMessage({ attachments: [] })
                ).rejects.toSatisfy((error: unknown) => {
                    const message = formatError(error);
                    expect(message.toLowerCase()).not.toContain("unhandled method");
                    expect(message.toLowerCase()).toContain("extension");
                    return true;
                });
            } finally {
                await session.disconnect();
            }
        }
    );
});

File: go/internal/e2e/rpc_server_misc_e2e_test.go
package e2e

import (
	"strings"
	"testing"
	"time"

	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
	"github.com/github/copilot-sdk/go/rpc"
)

func TestRpcServerMisc(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	sharedClient := ctx.NewClient()
	t.Cleanup(func() { sharedClient.ForceStop() })

	t.Run("should_reload_user_settings", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		if err := sharedClient.Start(t.Context()); err != nil {
			t.Fatalf("Start failed: %v", err)
		}

		if _, err := sharedClient.RPC.User.Settings().Reload(t.Context()); err != nil {
			t.Fatalf("User.Settings.Reload failed: %v", err)
		}
	})

	t.Run("should_report_agent_registry_spawn_gate_closed", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		client := newStartedIsolatedPortedClient(t, ctx)
		defer client.ForceStop()

		_, err := client.RPC.AgentRegistry.Spawn(t.Context(), &rpc.AgentRegistrySpawnRequest{Cwd: ctx.WorkDir})
		if err == nil {
			t.Fatal("Expected AgentRegistry.Spawn to be rejected by the closed spawn gate")
		}
		message := err.Error()
		assertPortedNoUnhandledMethod(t, message)
		assertPortedContainsFold(t, message, "agentRegistry.spawn")
		if !strings.Contains(strings.ToLower(message), "not enabled") && !strings.Contains(strings.ToLower(message), "no delegate") {
			t.Fatalf("Expected agentRegistry.spawn gate error, got %s", message)
		}
	})

	t.Run("should_shut_down_owned_runtime", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		client := newStartedPortedClient(t, ctx)
		defer client.ForceStop()

		if _, err := client.RPC.User.Settings().Reload(t.Context()); err != nil {
			t.Fatalf("User.Settings.Reload before shutdown failed: %v", err)
		}
		if _, err := client.RPC.Runtime.Shutdown(t.Context()); err != nil {
			t.Fatalf("Runtime.Shutdown failed: %v", err)
		}

		waitForRPCCondition(t, 15*time.Second, "runtime to stop serving RPCs after shutdown", func() (bool, error) {
			_, err := client.RPC.User.Settings().Reload(t.Context())
			return err != nil, nil
		})
	})

	t.Run("should_report_not_found_when_opening_session_without_context", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		client := newStartedIsolatedPortedClient(t, ctx)
		defer client.ForceStop()

		result, err := client.RPC.Sessions.Open(t.Context(), nil)
		if err != nil {
			t.Fatalf("Sessions.Open failed: %v", err)
		}
		if result.Status != rpc.SessionsOpenStatusNotFound {
			t.Fatalf("Expected Sessions.Open status not_found, got %+v", result)
		}
		if result.SessionID != nil {
			t.Fatalf("Expected nil session ID for not_found, got %q", *result.SessionID)
		}
	})

	t.Run("should_reject_send_attachments_from_non_extension_connection", func(t *testing.T) {
		ctx.ConfigureForTest(t)
		session := createPortedSession(t, sharedClient, nil)
		defer session.Disconnect()

		_, err := session.RPC.Extensions.SendAttachmentsToMessage(t.Context(), &rpc.SendAttachmentsToMessageParams{Attachments: []rpc.PushAttachment{}})
		if err == nil {
			t.Fatal("Expected SendAttachmentsToMessage from a normal SDK connection to fail")
		}
		message := err.Error()
		assertPortedNoUnhandledMethod(t, message)
		assertPortedContainsFold(t, message, "extension")
	})
}
